### ISSUE TITLE
(CRSI) Connors RSI

### ISSUE THREAD

**Issue by mrjbq7** [TA-Lib/ta-lib#431]
A request as made for a Connors RSI indicator, and code example provided

https://github.com/TA-Lib/ta-lib-python/issues/450

---

## Proposal

**Promote, scoped by oracle depth.** The roadmap's decline reason — *"no second
implementation"* (`to_be_deleted/ta-coverage-gap-and-roadmap.md:225`) — is **stale**,
but not for the reason `ta4j-catalog.md:110-121` gives.

- **trading-signals 8.3.0 reproduces the proposed formula across the parameter
  space.** MEASURED against a `TA_RSI`-exact reference on the 252-bar `ta_regtest`
  close series: max abs diff **≤ 2.13e-14** at `(3,2,100)`, `(3,2,5)`, `(14,14,20)`
  and `(2,2,2)` — and the same `outBegIdx` at all four. On a 400-bar random walk
  **≤ 2.84e-14**; on a 400-bar monotone ramp **0.000e+00**, bit-identical.
- **ta4j corroborates only at the published defaults.** A hand-composed ta4j
  expression is bit-identical to trading-signals at `(3,2,100)` (MEASURED
  0.000e+00 over all 151 outputs) and diverges by **18.34** index points at
  `(14,14,20)` and **9.82** at `(2,2,2)`, through `MMAIndicator`'s single-value
  Wilder seed. Its **rank leg alone is exact at every tuple** (MEASURED 0.0000).
- **ta4j's *shipped* `ConnorsRSIIndicator` computes a different formula.** It ranks
  the absolute close change, not the one-day percent return. MEASURED at the
  published defaults on the 252-bar series: differs from the canonical value on
  **123 of 151 output bars, max 3.3333** index points. `ta4j-catalog.md:114-121`
  called it "a real second implementation … not a reinterpretation" and asked for a
  numeric cross-check before promoting; the cross-check fails. Do not use that class
  as an oracle, and do not cite it as the second implementation.

The project's bar — *"an implementation outside this project to check it against"* —
is met. The scope is that goldens **outside** the published defaults rest on one arm.

## What it is

A 0–100 momentum oscillator published by Connors Research LLC (Laurence Connors,
Cesar Alvarez, Matt Radtke), *An Introduction to ConnorsRSI*, 2012, pp. 7–9. It
averages three short-term overbought/oversold measures with different mechanisms:
price momentum (a fast Wilder RSI), the duration of the current up/down close streak
(a very fast RSI of the signed run length), and the relative magnitude of today's
move (the percentile rank of today's one-day return against the previous 100). The
author's own equation:

> `ConnorsRSI(3,2,100) = [ RSI(Close,3) + RSI(Streak,2) + PercentRank(100) ] / 3`

Provenance is unusually clean: one named author, one dated primary document with
worked examples for both non-obvious pieces, no competing "original" formulation, and
every secondary source restating that equation with the same defaults.

## Formula

`m1 = optInTimePeriod` (3), `m2 = optInStreakPeriod` (2), `P = optInRankPeriod` (100).

**Component 1 — price momentum.** `TA_RSI(inReal, m1)`, unchanged: Wilder smoothing
with TA-Lib's SMA-of-the-first-`m1`-changes seed. No new numerics.

**Component 2 — the streak.** Signed run length of consecutive same-direction
closes; **an unchanged close resets it to 0**, it does not extend the run:

```
for i >= a+1:                              /* a = startIdx - lookbackTotal */
    d = inReal[i] - inReal[i-1]
    if      d == 0.0  streak[i] = 0
    else if d >  0.0  streak[i] = (streak[i-1] > 0) ? streak[i-1] + 1 :  1
    else              streak[i] = (streak[i-1] < 0) ? streak[i-1] - 1 : -1
```

MEASURED: reproduces the author's worked table (closes
20.00/20.50/20.75/19.75/19.50/19.35/19.35/19.40 → `_,1,2,-1,-2,-3,0,1`) exactly, and
is bit-identical to ta4j's `StreakIndicator` on **all 251 defined bars** of the
252-bar series (0 mismatches).

**Component 3 — the percentile rank.** Rank *today's one-day percent return* against
the **previous `P`** returns, **strictly less than**, denominator `P`:

```
ret[i]  = TA_ROCP(inReal, 1)              /* (c[i]-c[i-1])/c[i-1], 0.0 if c[i-1]==0 */
rank[i] = ( (double)count{ j in [i-P, i-1] : ret[j] < ret[i] } / (double)P ) * 100.0
          /* divide FIRST: `100.0*cnt/P` is a different double and is bitwise red
             against ta4j on 90 of 249 bars at P=3 (MEASURED, board card TA-Lib/ta-lib#368) */
```

The `×100` scaling of a return is irrelevant (a strictly increasing map preserves
rank order), so `TA_ROCP` — not `TA_ROC` — is the primitive, and its
`prevPrice == 0.0 → 0.0` guard (`ta_codegen/input/rocp/rocp.c:84-88`) comes with it.
Rank ∈ [0,100] and both RSI legs ∈ [0,100] by construction, and no other division
exists, so **no successful call can emit NaN or Inf** (#112 satisfied; the
`nan_inf_output` flag — whose users are `acos/asin/ln/log10/sqrt/div/vwma` — is
correctly not needed).

**Assembly:**

```
lookbackTotal = max( rsi_lookback(m1),        /* = m1 + unstable (classic compat) */
                     1 + rsi_lookback(m2),    /* streak is defined from bar a+1   */
                     1 + P );                 /* first full P-return window       */
for i = lookbackTotal .. endIdx:
    outReal[outIdx++] = ( rsiClose[i] + rsiStreak[i] + rank[i] ) / 3.0;
```

MEASURED: predicts the first emitted bar correctly at all four tuples checked against
trading-signals — `(3,2,100)→101`, `(3,2,5)→6`, `(14,14,20)→21`, `(2,2,2)→3` — and
agrees algebraically with its `getRequiredInputs()` (`max(m1+1, m2+2, P+2)` inputs).

**Load-bearing: the unstable-period term stays INSIDE the `max()` on both RSI legs.**
`max(rsi_lookback(m1), 1+rsi_lookback(m2), 1+P)`, never `max(m1, 1+m2, 1+P)`. MEASURED
at the defaults: dropping it is invisible up to unstable period 98, then the internal
RSI legs first emit **1 bar** late at 99 and **142 bars** late at 240 — and the range
sweep drives `TA_SetUnstablePeriod` to 240. CRSI would read values that were never
produced, on a sweep cell only.

**Anchoring: every leg seeds at `a = startIdx − lookbackTotal`**, the STOCHRSI shape
(`ta_codegen/input/stochrsi/stochrsi.c:101` calls `rsi(startIdx-lookbackSTOCHF, …)`).
This is not cosmetic — see Verification plan §1.

**Rejected variants:**

| Variant | Source | Why not |
|---|---|---|
| Rank the **absolute** change `c[i]-c[i-1]` | ta4j `ConnorsRSIIndicator` (wraps `DifferenceIndicator`, confirmed by `javap -c`) | Contradicts the primary source: *"we measure price change not in dollars and cents, but as a percentage of the previous day's price."* MEASURED: 123 of 151 bars, max 3.3333. |
| Rank denominator = count of valid window values | ta4j `PercentRankIndicator` (`getCountOfUnstableBars() = inner + period − 1`) | Emits one extra leading bar (MEASURED: index 100 at `(3,2,100)`, 20 at `(14,14,20)`) whose window holds `P−1` returns. Settled against by the author's worked example: *"Percent Rank = 3 / 20 = 0.15 = 15%"*. The arm must drop that bar. |
| Window `[i-P+1, i]` (includes the current bar) | the `PERCENTRANK` in the linked issue's sample | One bar short of *"the one day returns from each of the previous 20 days"*. |
| Rank the **raw price**; divide by `n` with no `×100` | same sample | Two further defects: it ranks `a[i]`, not the return, and returns 0..1 — so its `CONNORSRSI` averages a 0..1 term with two 0..100 terms. Its author states *"This is not verified."* Requesters will compare against this code; expect the report. |
| Streak extends through an unchanged close | none found | Contradicts the author's day-7 row. |
| Seed `RSI(streak,·)` at bar `a` with `streak[a]=0` | plausible transcription | MEASURED: invisible at the defaults (2.8e-14) but wrong by **0.3336** at `(14,14,20)` and **7.1429** at `(2,2,2)`. Anchor the streak RSI at bar `a+1`; ta4j and trading-signals both do. |
| `TA_MAType`-parameterised smoothers | the KC/VIDYA habit | Connors specifies Wilder RSI for both legs. |

## Proposed API

```c
TA_RetCode TA_CRSI( int    startIdx,
                    int    endIdx,
                    const double inReal[],
                    int    optInTimePeriod,     /* From 2 to 100000, default 3   */
                    int    optInStreakPeriod,   /* From 2 to 100000, default 2   */
                    int    optInRankPeriod,     /* From 2 to 10000, default 100  */
                    int   *outBegIdx,
                    int   *outNBElement,
                    double outReal[] );

int TA_CRSI_Lookback( int optInTimePeriod, int optInStreakPeriod, int optInRankPeriod );
/* max( rsi_lookback(m1), 1 + rsi_lookback(m2), 1 + optInRankPeriod ) */
```

- **Name `TA_CRSI`** — TradingView's own abbreviation. No collision: nothing matching
  `crsi`/`connors`/`streak`/`rank` among the 182 directories under
  `ta_codegen/input/`.
- Group `Momentum Indicators` (as `RSI`, `ta_codegen/input/rsi/rsi.yaml:2`); no new
  `TA_GroupId`. Input: plain `inReal`, single series (conventionally close). **One
  output** — well inside the 3-output cap.
- Defaults 3 / 2 / 100: the primary source's equation, restated by StockCharts,
  TradingView (`RSI Length 3`, `UpDown Length 2`, `ROC Length 100`), ta4j's no-arg
  constructor (bytecode pushes `iconst_3, iconst_2, bipush 100`) and
  trading-signals' constructor defaults (the StockCharts and TradingView pages are
  cited, not fetched this session; ta4j and trading-signals were read). `min = 2` on
  both RSI periods is `TA_RSI`'s
  own range and is **load-bearing**: `TA_RSI` with period 1 copies the input
  (`rsi.c:89-104`), which would leak raw prices into a 0–100 average. Hints use the
  #121 wording (`"Time period"`, …).
- **Flags: none.** Specifically:
  - **No `unstable_period`, and no new `TA_FUNC_UNST_CRSI`.** The lookback goes
    through `rsi_lookback()`, which already reads
    `TA_GetUnstablePeriod(TA_FUNC_UNST_RSI)` — exactly the STOCHRSI shape
    (`stochrsi.c:23`; `stochrsi.yaml:4` carries no `unstable_period`). A dedicated id
    would be **inert**, the #129 mistake that retired `TA_FUNC_UNST_STOCHRSI` to
    `UNUSED_22` (`ta_codegen/input/enums.yaml:74`), and it saves the seven hand edits
    a new id costs. It would also *actively* fail the xlang unstable-period axis,
    which requires the lookback to move between unstable 0 and 3
    (`src/tools/ta_regtest/CLAUDE.md:617-620`): at the defaults CRSI's is flat
    101→101 because `1+P` dominates.
  - **No `path_dependent`** — CRSI is *converging*, not path-dependent. See Notes.
- **Not a YAML-only change.** One row `{"CRSI", TA_FUNC_UNST_RSI}` is required in the
  `UNSTABLE_MAP` at `src/tools/ta_regtest/test_codegen.c:703`, next to
  `{"STOCHRSI", TA_FUNC_UNST_RSI}` and `{"ADXR", TA_FUNC_UNST_ADX}`. **Not** in the
  second copy at `server_verify.c:54`: that table is consulted only behind
  `fi->flags & TA_FUNC_FLG_UNST_PER` (`server_verify.c:486`) and its own comment
  (`:66-71`) says ADXR and STOCHRSI are omitted on purpose for exactly this reason —
  a CRSI row there would be dead code.

## References

**Fetched (this session):**

- **Primary.** Connors Research LLC, *An Introduction to ConnorsRSI* (2012), pp. 7–9
  — https://www.qmatix.com/ConnorsRSI-Pullbacks-Guidebook.pdf (fetched; text
  extracted locally, quoted verbatim here). Carries the streak table, the
  PercentRank definition (*"the number of values in the look-back period that are
  less than the current value, divided by the total number of values"*), the
  worked example (*"if the look back period is 20 days … three of those values are
  less than 2.0% … Percent Rank = 3 / 20 = 0.15 = 15%"*), the one-day-return
  definition (*"we measure price change not in dollars and cents, but as a
  percentage of the previous day's price"*), and the `(3,2,100)` equation.
- **USPTO TSDR** — https://tsdr.uspto.gov/statusview/sn85716238 (fetched).
  CONNORSRSI, Connors Research LLC, serial 85716238, filed 2012-08-29,
  Reg. No. 4397970, **Supplemental Register**, registered 2013-09-03, **status
  Dead/Cancelled — 2020-06-05**, "registrant did not file an acceptable declaration
  under Section 8".

**Cited, not fetched this session:**

- StockCharts ChartSchool — ConnorsRSI:
  https://chartschool.stockcharts.com/table-of-contents/technical-indicators-and-overlays/technical-indicators/connorsrsi
  — the page mrjbq7 linked in the request, at its pre-migration URL
  https://school.stockcharts.com/doku.php?id=technical_indicators:connorsrsi .
- TradingView — Connors RSI (CRSI):
  https://www.tradingview.com/support/solutions/43000502017-connors-rsi-crsi/

**Read from source on this machine:**

- ta4j 0.22.6 — `ConnorsRSIIndicator`, `helpers/{Streak,PercentRank,Difference}Indicator`,
  `RSIIndicator`, `averages/MMAIndicator`, disassembled with `javap -c` from
  Maven Central `org.ta4j:ta4j-core:0.22.6` (the version
  `ta4j_serve/pom.xml:34` pins).
- trading-signals 8.3.0 — `dist/momentum/CRSI/ConnorsRSI.js`, `momentum/RSI/RSI.js`
  under `trading_signals_serve/node_modules/`.
- The request and its unverified sample:
  https://github.com/TA-Lib/ta-lib-python/issues/450

## Sample input/output (board policy)

`TA_CRSI(TA_SREF_close_daily_ref_0_PRIV[252], 3, 2, 100)` —
`src/tools/ta_regtest/test_data.c:92`, the corpus `ta_regtest` already uses:

```
outBegIdx     = 101
outNBElement  = 151
outReal[0..5] = 58.138443061336396  25.292634245685434  65.63901198655614
                37.788267052246646  74.38994728880958   84.84436061248526
outReal[147..150] = 65.70687415488776  34.69246387648443
                    32.95592071145029  22.902915386547008
```

IEEE bits, big-endian hex (the form `capture.mjs` emits and the goldens should pin):
`404d11b880926623`, `40394aea13f2e898`, `405068e5928800ae`, … , `4036e7257678464c`.

Component breakdown at the first output bar (`i = 101`; `close[100] = close[101] =
116.00`):

```
RSI(close,3)   = 54.7751626679
streak         = 0                 <- unchanged close: the reset rule fires at outBegIdx
RSI(streak,2)  = 66.6401665161
PercentRank    = 53.0              <- 53 of the previous 100 returns are strictly smaller
CRSI           = (54.7751626679 + 66.6401665161 + 53.0) / 3 = 58.1384430613
```

MEASURED: trading-signals 8.3.0 `ConnorsRSI` reproduces the whole 151-value vector to
**1.421e-14 absolute / 5.931e-16 relative**; the hand-composed ta4j expression is
**bit-identical to trading-signals** on all 151 values at this tuple.

## Oracles

| Oracle | Covers CRSI | Arm status | Work to add the arm |
|---|---|---|---|
| `trading_signals_serve` | **yes, across the parameter space** — `momentum/CRSI/ConnorsRSI`, matches the primary source | **must be added** | one `ARMS` row in `capture.mjs` (three arms today: AC, AO, VWAP) plus a third `feed` mode: every existing arm feeds a candle object (`'hl'` / `'ohlcv'`), but `ConnorsRSI.update()` takes a bare `number`, so `bar()` needs a `'close'` case. The constructor takes an options object: `ctor: (p) => new ts.ConnorsRSI({rsiInterval:p[0], streakRsiInterval:p[1], percentRankInterval:p[2]})`. `ConnorsRSI` is exported from the package root (via `dist/momentum/index.js:10`). ~5 lines. |
| `ta4j_serve` | **defaults only, and only hand-composed** | **optional** | one `Server.Arm` + one `ARMS.put(...)` (`Server.java:122,148`). Body must be `[ RSIIndicator(close,m1) + RSIIndicator(StreakIndicator(close),m2) + PercentRankIndicator(new ROCIndicator(close,1), P) ] / 3` — **not** `new ConnorsRSIIndicator(...)` — and must discard the extra leading bar. Capture at `(3,2,100)` only. |
| `ta_tulip_serve` | no | n/a | Tulip 0.9.2 has no ConnorsRSI, no streak and no rank/percentile indicator (no matching file in `tulip_serve/vendor/tulipindicators/indicators/`). |
| `ta_pandas_serve` | no | n/a | pandas-ta-classic 0.6.52 has no `crsi.py`; a package-wide grep for `connors`/`crsi`/`percent_rank`/`percentrank` returns nothing. |

MEASURED facts a future arm author needs:

- **The ta4j arm is a defaults-only oracle.** Whole-vector max abs diff against a
  `TA_RSI`-exact reference on the 252-bar series:

  | tuple | ta4j composed | trading-signals |
  |---|---|---|
  | (3,2,100) | 1.421e-14 | 1.421e-14 |
  | (14,14,20) | **1.834e+01** | 2.132e-14 |
  | (2,2,2) | **9.820e+00** | 1.421e-14 |

  Per leg at `(14,14,20)`: `RSI(close)` off **23.1830**, `RSI(streak)` off
  **31.8293**, rank leg **0.0000**. At `(2,2,2)`: `RSI(close)` off **29.4614**, the
  other two 0.0000. Cause: `org.ta4j.core.indicators.averages.MMAIndicator` seeds
  Wilder with the **single first value** where `TA_RSI` seeds with the SMA of the
  first `period` changes (`rsi.c:189-214`). At `(3,2,100)` the 98 spare warm-up bars
  inside `1+P` decay that below double rounding; at small `1+P` they do not. **The
  rank leg generalises; the RSI legs do not.**
- **ta4j's streak anchoring matches the proposal.** `StreakIndicator.getCountOfUnstableBars()`
  returns 1, and MEASURED `RSIIndicator(StreakIndicator, m2)` first emits at index
  `1+m2` at all three tuples. None of the 18.34 gap is anchoring.
- **trading-signals diverges on a flat series.** Its `RSI` returns `100` whenever
  `avgLoss === 0`, including when `avgGain === 0` too, where `TA_RSI` returns `0.0`
  — the explicit #253 decision at `ta_codegen/input/rsi/rsi.c:226-233`. MEASURED on a
  400-bar constant series at all three tuples: trading-signals `66.6666666667`,
  canonical `0.0` — **exactly 66.667 apart**. Exclude constant and near-constant
  segments from that arm's corpus, or the gate reds on a value TA-Lib is right about.
  ta4j agrees with TA-Lib here.
- **Random-walk and ramp corpora**, so no claim here rests on one series. MEASURED
  against trading-signals, 400 bars each, all three tuples: random walk **1.42e-14 /
  1.42e-14 / 2.84e-14**; monotone ramp **0.000e+00** with the value exactly
  `66.6666666667` (both RSI legs saturate at 100, every new return is smaller than
  all `P` predecessors so the rank is 0) — the ramp is where trading-signals' `100`
  and TA-Lib's `100` coincide, so it is *not* a substitute for the flat-series test.
- Both arms are value-only; the ta4j arm must publish `beg = 1 + P`.

## Verification plan

1. **Composite differential — a policy deviation, stated as one.**
   `to_be_deleted/composite-functions.md` §2 disqualifies a decomposition needing a
   primitive we do not ship — *"the reference would be novel unverified code"*. CRSI
   needs two: the streak recurrence and the percent-rank loop. Board card **#17 supplies
   the second one exactly** — its `TA_PERCENTRANK(inReal, n)` (window `[i-n, i-1]`, strict
   `<`, denominator `n`, lookback `n`) is this component applied to `TA_ROCP` rather than
   to `close`. Its companion `TA_PERCENTILE` is the *inverse* question and is not used
   here. The streak recurrence stays novel either way, so CRSI is outside §2 with or
   without TA-Lib/ta-lib#368 — TA-Lib/ta-lib#368 is a supplier, not a blocker. The deviation is defensible but a human owes the call
   (Open questions): the novel code is **integer-exact** (a `{0,±1,±n}` state machine
   and an integer count over an integer denominator), all the floating-point lives in
   the two `TA_RSI` calls, and both novel pieces have an *external* check — the
   streak against the author's table and ta4j's `StreakIndicator` (MEASURED 0
   mismatches on 251 bars), the rank against ta4j's `PercentRankIndicator` over
   `ROCIndicator(close,1)` (MEASURED 0.0000 at every tuple).

   **The anchor is the whole test.** Anchor every leg at CRSI's own
   `a = outBegIdx − TA_CRSI_Lookback(...)`: call `TA_RSI(close, m1)` with
   `sr = a + rsi_lookback(m1)`, reading bar `i` at output offset `i − sr`; build the
   streak buffer from bar `a` and call `TA_RSI(streak, m2)` at buffer index
   `rsi_lookback(m2)`. This is the SMI
   precedent verbatim (`src/tools/ta_regtest/ta_test_func/test_composite2.c:62-68`:
   *"for a requested startIdx it begins its TA_MAX/TA_MIN at (refStart - lookback) +
   (period-1), so each TA_EMA seeds on exactly the bars the fused loop seeds on"*).
   MEASURED, startIdx swept in steps of 7 over the 252-bar series: that anchor is
   **bit-exact (0.000e+00) at all four tuples**, while the two plausible wrong
   anchors are not — `startIdx = outBegIdx` gives `CRSI[101] = 54.7746705794` against
   the published `58.1384430613` (**3.3638** off at the first bar), and
   `startIdx = a` (which `TA_RSI` re-clamps to its own lookback) is off by up to
   **12.58** index points at `(2,2,2)`.

   Assert **bit-exact** (`memcmp`) on that anchor: both sides feed `TA_RSI` identical
   `double`s and the rank is an integer count over an integer. Grid `(3,2,100)`,
   `(3,2,5)`, `(14,14,20)`, `(2,2,2)`, each at a sweep of `startIdx` and at
   `startIdx == endIdx == lookback` — **crossed with an unstable-period sweep**,
   which is the load-bearing half: the `max()` invariant above is invisible at
   unstable period 0 and at every tuple where `1+P` dominates.

   **New file, and it does churn the source lists.** `test_composite.c` no longer
   exists; the files are `test_composite1.c` (nine indicators) and
   `test_composite2.c` (SMI), and `test_composite2.c:53-58` states the convention —
   *"every other group in `ta_test_func/` is one file per group"*. So CRSI gets
   `src/tools/ta_regtest/ta_test_func/test_crsi.c`, registered in **both**
   `CMakeLists.txt` (next to `:660-661`) and `src/tools/ta_regtest/Makefile.am`
   (next to `:50-51`); `scripts/build.py check-source-lists` is the check, and the
   autotools entry is what the dist nightly builds.
2. **External goldens — the load-bearing half.** The differential proves the
   optimization, not the formula: a wrong streak rule sits on both sides of it.
   Freeze the 151-value `(3,2,100)` vector above as hardcoded expected values **as
   IEEE hex bits**, not decimals — `capture.mjs` emits both, and only the hex
   round-trips. Compare at relative tolerance **1e-12** with a **1e-12 absolute
   floor** (the PVO precedent, `test_composite1.c:162-169`); the floor is not
   ornamental, because CRSI is exactly `0.0` on a flat segment where a relative-only
   tolerance is undefined. That golden may be **dual-sourced**: trading-signals and
   the hand-composed ta4j arm are bit-identical at this tuple. Add a second capture
   at `(14,14,20)` from **trading-signals alone** — it is where the ta4j arm stops
   being usable, and it is also the tuple that separates the correct streak anchor
   from the plausible wrong one (MEASURED 0.3336 apart).
3. **Free coverage.** Once the YAML lands, `TA_ForEachFunc` picks CRSI up for
   `--codegen` across all four backends, `--xlang-hash` / `server_verify` at zero
   tolerance, `--fuzz-064`, the #94 boundary sweep under ASan/UBSan, and #108
   output-aliasing rejection.
4. **Range stability.** With the `UNSTABLE_MAP` row, `stability_class()`
   (`test_codegen.c:1878`) answers `TA_STABLE_CONVERGING` and the sweep warms
   `TA_FUNC_UNST_RSI` across 0..240. Any hand-written range test must pass
   `TA_FUNC_UNST_RSI` — `doRangeTestMulti`'s guard (`test_util.c:614-637`) rejects
   `CONVERGING` with `TA_TEST_UNST_NONE`, and rejects `EPSILON` *with* an id. This is
   not cosmetic: MEASURED range dependence (`startIdx` swept in steps of 7, compared
   against the call at the lookback) is **7.294** at `(3,2,5)`, **2.908** at
   `(14,14,20)` and **16.64** at `(2,2,2)` index points, against `TA_STABLE_EPSILON`'s
   1e-10 absolute tolerance (`test_util.c:558-566`). It is **0.000e+00** at the
   defaults (at unstable period 0, 3, 20 *and* 140), so the generic codegen range
   gate — which runs at default parameters after `reset_opt_periods_to_default` —
   would pass either way. That is the PVO trap exactly (`test_codegen.c:741-744`:
   *"Its absence classified it EPSILON … PVO moves by whole multiples across
   startIdx"*): green, and wrong.
5. **Hand-written edge cases:**
   - **Constant input** — assert exactly `0.0`. Exercises `TA_RSI`'s `gain+loss == 0`
     arm on both legs *and* the all-ties rank. This is the case that catches a
     transcription copied from trading-signals rather than TA-Lib (MEASURED 66.667
     apart).
   - **A zero close** — assert finite output. MEASURED with the `TA_ROCP` guard: every
     output finite. MEASURED without it: a single NaN return silently corrupts the
     rank on **50 of the following 100 bars** and **never produces a NaN output**,
     because `NaN < x` is false and the count merely comes out low. Silent, so it
     earns the test.
   - **Unchanged close** — assert the streak resets and does not extend. The standard
     corpus exercises this at `outBegIdx` itself (bars 100 and 101 both close 116.00);
     pin it explicitly.
   - **`optInRankPeriod` small enough that an RSI leg dominates** — `(14,14,20)` and
     `(2,2,2)`; assert `outBegIdx` against the three-way `max()`.
   - `nbBars < lookback` → `TA_SUCCESS`, `outNBElement == 0`, `outBegIdx == 0`.
   - **In-place aliasing** (`inReal == outReal`): the streak, the previous close and
     the return ring all read `inReal` behind the write cursor, so this is safe only
     if they live in state and are never re-read from `inReal`. ASan will not catch
     it.
   - **Monotone ramp** — both RSI legs saturate at 100, the rank is 0, the value is
     exactly `66.6666666667` at every tuple (MEASURED, and bit-identical to
     trading-signals). Assert no NaN.

## Open questions

1. **`composite-functions.md` §2 deviation.** §2 disqualifies a decomposition that
   needs an unshipped primitive; CRSI needs two. Verification plan §1 argues the
   deviation is bounded (both novel pieces are integer-exact and both are externally
   checked) — but that is a policy call a human owes, not a settled matter. The
   alternative is to ship CRSI with **no** differential leg and rely on the external
   goldens plus the range/xlang gates alone.
2. **`TA_COMPATIBILITY_METASTOCK`.** As written, both `rsi_lookback()` calls inherit
   the `−1` (`rsi.c:20-30`) *and* `rsi()` takes a different seeding branch
   (`rsi.c:119-185`, the synthetic extra first bar), so a METASTOCK-mode CRSI changes
   **values**, not just the lookback. MEASURED on the 252-bar series: `outBegIdx` is
   unchanged at all three tuples, and the values differ by **0.0000** at `(3,2,100)`,
   **0.2866** at `(14,14,20)` and **4.2327** at `(2,2,2)`. The precedent answer is
   "inherit, like STOCHRSI"; nobody has confirmed that a "Metastock ConnorsRSI" is a
   thing anyone wants, and no source defines one.
3. **`optInRankPeriod` minimum.** Proposed **2**, to match both RSI legs and because
   `P = 1` makes the rank a two-valued `{0, 100}` sign test rather than a rank. No
   published source constrains it; ta4j requires `≥ 1`, trading-signals validates
   nothing. Decide before the first release — the range is ABI.
4. **Whether to ship `[stream]` in the first release.** See Notes; it is a cost
   decision with a measurement attached, not a shape question.

**Settled, not open:** the trademark. `CONNORSRSI` (Connors Research LLC, serial
85716238, Reg. No. 4397970, Supplemental Register) is **DEAD — cancelled 2020-06-05**
for failure to file a Section 8 declaration (FETCHED from TSDR). No live US
registration blocks the name, and `#55 TDS` is not a precedent: it is **open**,
*"proposing No Status (HOLD)"*, for a **scope** disagreement between two
implementations, with trademark an unresolved question *inside* that card. What is
live is copyright in the guidebook, which states *"This document cannot be reproduced
without the expressed written permission of Connors Research, LLC"* — ta-lib.org may
state the formula (an idea) but must not reproduce the guidebook's prose or its
worked table.

## Notes

- **Not start-dependent (#127 / `TA_FUNC_FLG_PATH_DEP`) — CRSI is CONVERGING.** The
  streak looks like a path-dependent state machine, but it is not the source of range
  dependence: the streak reaches `TA_RSI` only through a difference, so a wrong
  re-seed perturbs at most the first direction flip after the anchor. The dependence
  that exists lives in the two **Wilder recurrences** — the same shape as STOCHRSI
  and ADXR — and decays geometrically. MEASURED range dependence: **0.000e+00** at
  `(3,2,100)`, **7.294 / 2.908 / 16.64** at `(3,2,5)` / `(14,14,20)` / `(2,2,2)`. The
  zero at the defaults is not a property of CRSI, it is `1+P` dominating the `max()`:
  the `m1 = 3` leg gets 98 spare warm-up bars and `(2/3)^98` is below double rounding.
  Setting `path_dependent` would push CRSI into `TA_STABLE_SKIP`, where the range test
  compares **nothing** (`src/tools/ta_regtest/CLAUDE.md:119`) — lost coverage for no
  gain.
- **Batch complexity is O(n·P), and that is a shipped-body decision.** With
  `optInRankPeriod` capped at 10000, as PERCENTRANK is for the same per-bar count, a
  naive count is still the dominant cost.
  MEASURED on this box (AMD Ryzen 7 PRO 8840U, WSL2, gcc -O2, 200k-bar random walk,
  best of 3): the rank leg alone costs **9.05 / 46.74 / 461.65 ns/bar** at
  `P = 20 / 100 / 1000`, against **1.29 ns/bar** for an `SMA(100)` running sum on the
  same corpus — ~36x at the default `P`, and linear in `P`. The in-tree precedent for
  shipping the naive loop is `AVGDEV`, which runs *two* O(period) inner loops per bar
  (`ta_codegen/input/avgdev/avgdev.c:51,56`). The shipping card must state whether
  the body is that naive loop; an incremental rolling-rank structure is a separate
  optimization with its own differential.
- **Streaming: deferred, untested, and NOT blocked by a census rejection.** Four of
  five pieces stream trivially (two `TA_RSI` states, two streak scalars, one previous
  close). The fifth is a `P`-deep ring re-scanned per bar — the same shape `AVGDEV`
  already streams with (`avgdev.yaml:4` carries `[overlap, stream]`), so this is a
  **cost** question, not a shape one. The DONCHIAN precedent points the other way from
  what it is usually cited for: `donchian.yaml:4` carries `[overlap, stream]` today,
  and what `stream-census` rejected there was a second window *cursor*, which was
  removed. What is genuinely untested here is whether the emitter accepts a rescan
  that accumulates a **comparison** (`cnt += ret[j] < ret[i]`) rather than a sum.
  Ship batch first, revisit like PVO.
- **ABI: there is no C `TA_FuncID`.** Nothing in `include/` defines one; the C side is
  name-keyed tables plus `src/ta_abstract/ta_group_idx.c`. The identifier that
  renumbers is the Rust crate's `#[repr(u16)] pub enum FuncId`
  (`ta_codegen/output/rust/library/src/abstract_api.rs:44`), a dense alphabetical
  index of 181 entries — `CRSI` inserts between `COSH` (`:222`) and `DEMA` (`:224`)
  and shifts every later discriminant. That is a Rust-crate semver note, and it
  applies **because** CRSI sorts mid-table. The separate `#228` `ta_ref_serve`
  head-of-table hazard does **not** apply for the same reason; verify anyway with
  `bin/ta_abstract_dump_new` vs `_ref`.
- Release chores that have bitten this repo: run a **full** `cargo run -- generate`
  (a `--func=CRSI` filter skips `Core.java` and the servers), then
  `scripts/build.py servers` — not `build`, or the gate coverage reads green while
  stale; and add the `CHANGELOG.md` entry under `[0.8.1] … New TA Functions`, which
  was forgotten for CMOU.

**Comment by mario4tier** [comment]
## Ruling: ships with `flags: [stream]`

Generator work is accepted **provided it is general purpose** — a composed-plan capability other functions
can reach, not a CRSI branch.

Scope of that work, stated precisely so it is not over- or under-estimated. Two separate things were being
conflated on this card:

- **The batch body needs no generator work at all.** It is STOCHRSI's composed shape with three sub-calls
  instead of one: `rsi()` twice, `rocp()`, `percentrank()`, then a three-way average. Every construct is
  already emitted for a shipped function.
- **The streaming lowering is where the new capability is.** Of the 11 shipped composed plans, only
  STOCH/STOCHF have a producer loop and both have a single-sub-call tail; CRSI is a producer plus three
  sub-calls plus an offset combine. The sub-stream embedding it needs already exists
  (`TA_RSI_Stream *sub0` inside `struct TA_STOCHRSI_Stream`, `src/ta_func/ta_STOCHRSI.c:304`; APO/PPO/PVO/
  STOCH each embed two, MACDEXT three), and `TA_PERCENTRANK` already ships `flags: [stream]` with a
  P-deep ring rescanned by a comparison accumulator (`src/ta_func/ta_PERCENTRANK.c:224-250`).
- **Known trap on that path:** the streak recurrence `s = (s > 0) ? s + 1 : 1` is not the bare `v++` the
  paced-variable detector scans for. Expect that to be where the lowering refuses, and fix it in the
  detector rather than by reshaping the streak.

First step is one `generate` run in a private tree to find out what `analyze_composed` actually rejects.
That is an hour, and it is what separates "an hour" from "a day" on the rest.

## What changed since this card was written

`TA_PERCENTRANK` landed (TA-Lib/ta-lib#369) with exactly the semantics the author specifies — window
`[i-P, i-1]` excluding today, strictly `<`, denominator `P`, divide-then-scale. **The rank leg is no longer
novel code**, so the only new arithmetic in the whole function is the 8-line integer streak state machine.
`percentrank.yaml`'s `range: [2, 10000]` also answers this card's own open question 3, and sets
`optInRankPeriod`'s top: the count grows with the period, which is why PERCENTRANK stops at 10000.

## Primary source — read, not cited

Connors / Alvarez / Radtke, *An Introduction to ConnorsRSI* (Connors Research, 2012), §1 pp. 7-9. All four
CG-trap-prone pieces are stated verbatim by the authors:

1. **Unchanged close resets the streak to 0** — their 8-row worked table (closes 20.00 / 20.50 / 20.75 /
   19.75 / 19.50 / 19.35 / 19.35 / 19.40 giving `_, 1, 2, -1, -2, -3, 0, 1`) plus "On Day 7 the closing
   price is unchanged, so the Streak Duration is set to 0". Verified by direct computation against all
   seven defined rows.
2. **The rank is over the one-day PERCENT return** — "not in dollars and cents, but as a percentage of the
   previous day's price".
3. **Strictly less than, denominator P, window excludes today** — "the number of values in the look-back
   period that are less than the current value, divided by the total number of values", worked as
   `3 / 20 = 0.15`.
4. **Assembly and defaults** — `[ RSI(Close,3) + RSI(Streak,2) + PercentRank(100) ] / 3`.

TradingView's own support page corroborates the strict `<` independently.

## Oracles

| Oracle | Status | Work |
|---|---|---|
| trading-signals 8.3.0 `ConnorsRSI` (TS, MIT) | **executed this session** | ONE line in `capture.mjs` |
| FacioQuo published 502-row streak column | published golden | transcription, zero code |
| ta4j `ConnorsRSIIndicator` | **REFUSE — do not use** | — |

trading-signals vs a naive-compose reference built from the current shipped input C, on the 252-bar
`TA_SREF_close_daily_ref_0_PRIV` series: `(3,2,100)` beg 101 nb 151 max abs **1.421e-14**;
`(3,2,5)` beg 6 nb 246 max abs 1.421e-14; `(14,14,20)` beg 21 nb 231 max abs 2.132e-14;
`(2,2,2)` beg 3 nb 249 max abs 1.421e-14. `outBegIdx` identical at all four.

**ta4j must be refused as an oracle here**, and this is the CG trap in the open: its
`ConnorsRSIIndicator` ranks the *absolute* close change via `DifferenceIndicator` and divides by the count
of valid values — it diverges at two of the three places the author is specific. A card resting on
"we agree with ta4j" would have shipped the wrong indicator.

The card's note that `capture.mjs` needs a close-feed mode added is **stale**: it has ~13 arms and
`feed: 'close'` is already used by VHF, FOSC, TSI and RVI.

## Two items settled by measurement — record, do not re-litigate

- **The streak is NOT `path_dependent`.** It re-synchronises at the first direction flip; the measured
  start-dependence residual decays from 16.6 to 2.5e-4 index points as the unstable period goes 0 -> 20.
  It is CONVERGING, and needs the row `{"CRSI", TA_FUNC_UNST_RSI}` in `UNSTABLE_MAP`
  (`src/tools/ta_regtest/test_codegen.c:703`, next to STOCHRSI). Classifying it EPSILON would be green and
  wrong — the generic codegen range gate runs at DEFAULT parameters, where the deviation is exactly zero.
- **Lookback is a three-way max with the unstable term INSIDE it:**
  `max( rsi_lookback(m1), 1 + rsi_lookback(m2), 1 + P )`. The naive `max(m1, 1+m2, 1+P)` already diverges
  at unstable period 1 for `(2,2,2)`. No new `TA_FUNC_UNST_*` id — CRSI inherits RSI's, the STOCHRSI shape.

## Implementation trap, measured

Do **not** prepend a synthetic `streak[a] = 0` and anchor the streak RSI at bar `a`. That variant is
invisible at the defaults (0.0 difference) but wrong by 0.694 at `(3,2,5)`, 0.334 at `(14,14,20)` and
7.143 at `(2,2,2)` index points — **with an identical `outBegIdx`**, so no range or `outBegIdx` check can
catch it. The goldens therefore need at least two parameter tuples.

Related: a transcription copied from trading-signals gives 66.667 on a flat series where TA-Lib must give
exactly 0.0 — its RSI returns 100 when `avgLoss == 0` even if `avgGain == 0`, where `TA_RSI` returns 0.0
(TA-Lib/ta-lib#253).

## Remaining choice — implementation, not a blocker

**Composed body or fused body?** Recommendation: **composed**. It makes the shipped body its own
naive-compose reference, so no differential leg against unshipped logic is needed and
`to_be_deleted/composite-functions.md` §2 never applies. The performance argument for fusing is weak:
measured on this box, the shipped `percentrank.c` alone is 14.88 / 66.59 / 717.68 ns/bar at P = 20 / 100 /
1000, so the rank rescan is 41% of a composed CRSI at the default P=100 and 86% at P=1000. Fusing the
cheap legs buys little; an incremental rolling-rank structure is the real optimization and is separate
work with its own differential.

Effort **M**. Trademark: dead. Demand: ta-lib-python#450, a named requester.

**Comment by mario4tier** [comment]
I am taking this one.

**Comment by mario4tier** [comment]
## Shipped on dev as `c9e67b0fb`

Composed body as ruled: `rsi()` twice, `rocp()`, `percentrank()`, then the three-way average, with `flags: [stream]`, no new `TA_FUNC_UNST_*`, and the unstable term inside the lookback's `max()`. Ranges 3/2/100 over [2,100000] / [2,100000] / [2,10000], the last following `percentrank.yaml`.

**The streaming lowering needed one composed-plan capability, and it was not the predicted one.** The paced-variable detector accepted the streak recurrence as written; nothing was reshaped for it. What `analyze_composed` refused was the combine map: it reads two sub-outputs at an offset, and those sub-calls run over different index spaces, which the same-bar proof could see only when both spelled the same `endIdx` argument. A series now also carries whether its last element is the `endIdx` bar, seeded from the producer loop and carried through chained sub-calls; an offset read is same-bar when the offset is the element-count difference and both series end there. General, not a CRSI branch, and it leaves every shipped function's generated output byte-identical.

**Body layout.** Three buffers, which is the minimum while the output may alias the input: the streak RSI writes its own buffer, `rocp` reuses the dead streak buffer, `percentrank` writes `outReal`. Writing the streak RSI in place instead cost Rust a fourth allocation plus a `mem::swap`, and cost every backend's `Open` a second RSI pass.

## Two corrections to this card

- **The sample vector above is TA_CRSI's own output, not trading-signals'.** The last value differs by 1 ulp (`4036e7257678464c` here, `...464b` from the library). The goldens were frozen from the executed capture instead, never from the card.
- **Open question 2 is moot:** `TA_COMPATIBILITY_METASTOCK` was retired in #388, so RSI has no Metastock branch left to inherit and lookbacks and values are identical in both modes. Question 1 falls away with the composed ruling (the body is its own naive-compose reference), 3 is answered by PERCENTRANK's own range, and 4 shipped as `[stream]`.

## Verification

Goldens are trading-signals 8.3.0, executed and frozen as IEEE bits over four series: the 252-bar corpus at (3,2,100), (14,14,20) and (2,2,2); a tick series; a 20-bar streak series with unchanged closes after runs of both signs; and one anchored row. `ta-lib-oracles/capture_431_crsi.py` reproduces all of them in a single run and cross-checks an exact rational reference of the definition; TA_CRSI is within 1.42e-14 absolute of exact on every series, tuple, anchor and unstable period measured.

The 252-bar corpus alone cannot separate a strict rank from a tie-counting one, because no window in it holds a tied return; that is what the tick and streak rows are for. The zero-close windows come from the exact arm, trading-signals having no previous-close == 0 guard and being up to 17 index points off there.

An adversarial review found two test gaps, both now closed and each shown to fail on its mutant: the anchor leg ran only at unstable period 0, where dropping the unstable term from the close leg's offset is invisible, and the zero-close leg asserted only finiteness, which holds with or without the ROCP guard.

Gates: `ta_regtest` (CRSI and full), `--codegen` per-function and unfiltered, `--xlang-hash` bit-identical across Rust, Java and C#, the Java jar plus the C# suites, Rust doc and lib tests, `cargo doc` warning-free, clippy on both crates, the generator suite, `check-source-lists`, `regen-check`, and a second `generate` leaves the tree unchanged.

## Measured, and left undone

- **Cost is 46 / 50 / 216 ns/bar at P = 20 / 100 / 1000** (Ryzen 7 PRO 8840U, 200k-bar walk). PERCENTRANK is about half of it at the default P and 88% at P=1000, as this card predicted.
- **The streak loop is about 21% at P=100**, all branch mispredicts. A branch-free integer form is 4x on that loop and bit-identical, but the only spelling gcc keeps branch-free needs generator work in two backends: C# and Java collapse `cond ? 1 : 0` to a bool inside arithmetic, and Rust types the `(a>b?1:0)-(a<b?1:0)` difference as `usize`. Every other spelling gets re-branched by gcc. Not done.
- **A sorted-window PERCENTRANK loses at P=100** and only pays off near P=10000, so the shipped body stays the naive count.
